torch/csrc/distributed/rpc/python_functions.cpp

```diff
@@ -109,52 +110,124 @@ std::shared_ptr<FutureMessage> sendPythonRemoteCall(
 
 using namespace torch::distributed::autograd;
 
-std::shared_ptr<FutureIValue> toFutureIValue(
-    const std::shared_ptr<FutureMessage>& fm,
+c10::intrusive_ptr<JitFuture> wrapFutureMessageInJitFuture(
+    const std::shared_ptr<FutureMessage>& futureResponseMessage,
     bool hasValue) {
-  auto fv = std::make_shared<FutureIValue>();
-
-  fm->addCallback([fv, hasValue](const FutureMessage& fm) {
-    if (fm.hasError()) {
-      fv->setError(*fm.error());
-    } else if (hasValue) {
-      // Don't need to acquire GIL here, as toPyObj and toIValue acquires GIL
-      // when creating/copying the py::object
-      fv->markCompleted(
-          jit::toIValue(toPyObj(fm.constValue()), PyObjectType::get()));
-    } else {
-      fv->markCompleted(IValue());
-    }
-  });
-
-  return fv;
+  if (hasValue) {
+    // NB: The custom deleter is necessary because the FutureIValue object
+    // holds a py::object and it would require GIL to delete.
+    c10::intrusive_ptr<JitFuture> jitFuture =
+        c10::make_intrusive<JitFuture>(PyObjectType::get());
+
+    futureResponseMessage->addCallback(
+        [jitFuture](const FutureMessage& futureResponseMessage) {
+          // Don't need to acquire GIL here, as toPyObj acquires GIL
+          // when creating the py::object
+          if (futureResponseMessage.hasError()) {
+            jitFuture->setError(futureResponseMessage.error()->what());
+          } else {
+            jitFuture->markCompleted(jit::toIValue(
+                toPyObj(futureResponseMessage.constValue()),
+                PyObjectType::get()));
+          }
+        });
+
+    return jitFuture;
+  } else {
+    c10::intrusive_ptr<JitFuture> jitFuture =
+        c10::make_intrusive<JitFuture>(NoneType::get());
+
+    futureResponseMessage->addCallback(
+        [jitFuture](const FutureMessage& futureResponseMessage) {
+          if (futureResponseMessage.hasError()) {
+            jitFuture->setError(futureResponseMessage.error()->what());
+          } else {
+            jitFuture->markCompleted(IValue());
+          }
+        });
+
+    return jitFuture;
+  }
 }
 
-std::shared_ptr<FutureIValue> pyRpcBuiltin(
+c10::intrusive_ptr<JitFuture> pyRpcBuiltin(
     const WorkerInfo& dst,
     const std::string& opName,
+    const float rpcTimeoutSeconds,
```

Comment:
This is creating discrepency between other APIs, `pyRpcTorchscript` puts `rpcTimeoutSeconds` as the last arg and this put in the middle, we should make this also as the last arg.

Reply:
That will not work. `py::kwargs` captures all rest kwargs. `rpcTimeoutSeconds` will be captured into `kwargs` for user function, which is not the expected behavior.

Reply:
Updating.